## Step : 0 Collecting Data :- https://www.kaggle.com/datasets/ahmedmohamed2003/retail-store-sales-dirty-for-data-cleaning  

## Step : 1 Import Required Python Library / Load Dataset

### Import Required Library

In [2]:
# for inside data
import numpy as np

# for dataframe 
import pandas as pd

# for visulation
import matplotlib.pyplot as plt

import seaborn as sns

In [3]:
import warnings
warnings.filterwarnings('ignore')

### Load Dataset

In [10]:
df = pd.read_csv('Data/retail_store_sales.csv')
df.head()

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,NaN
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False


### Step :- 2 Basic Information

#### i) Columns names

In [11]:
df.columns

Index(['Transaction ID', 'Customer ID', 'Category', 'Item', 'Price Per Unit',
       'Quantity', 'Total Spent', 'Payment Method', 'Location',
       'Transaction Date', 'Discount Applied'],
      dtype='object')

#### ii) Check duplicate rows

In [12]:
df.duplicated().sum()

0

#### iii) Check Numbers of Rows and Columns

In [16]:
print(f"Number of Rows    :- {df.shape[0]}")
print(f"Number of Columns :- {df.shape[1]}")

Number of Rows    :- 12575
Number of Columns :- 11


#### iv) Check for Null values

In [14]:
df.isnull().sum()

Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit       609
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64

#### v) Overall Information

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


#### Interpretation
- This dataset contains **12575 observations(Rows)** and **11 Attributes(Columns)**
- We have `3 Numerical Columns` and `8 Categorical Columns`
- This Datasets `Not contain Duplicate Values`
- Missing values are present in the 
    - **Item** (1213 records)
    - **Price Per Unit** (609 records)
    - **Quantity** (604 records)
    - **Total Spent**  (604 records)
    - **Discount Applied** (4199 records) columns of the dataset.

### Step :- 3 Data Preprocessing / Data Cleaning

### Null Value Handling
- There are two types of null values in a dataset:

- Valid null value: Represents implicit missing records

- Invalid null value: Represents explicit missing records

### Guidelines for Handling Null Values
- If the percentage of missing values is less than 10%, remove the observations.

- If the percentage of missing values is between 20% and 70%, impute the missing records.

- If the percentage of missing values is more than 80%, remove the attribute (column) entirely.

In [18]:
(df.isnull().sum()/df.shape[0]) * 100

Transaction ID       0.000000
Customer ID          0.000000
Category             0.000000
Item                 9.646123
Price Per Unit       4.842942
Quantity             4.803181
Total Spent          4.803181
Payment Method       0.000000
Location             0.000000
Transaction Date     0.000000
Discount Applied    33.391650
dtype: float64

#### Filling Missing Values in Discount Applied

In [34]:
df['Discount Applied'].value_counts()

Discount Applied
True     4219
False    4157
Name: count, dtype: int64

In [35]:
df['Discount Applied'].mode()[0]

True

In [37]:
df['Discount Applied'] = df['Discount Applied'].fillna(value=df['Discount Applied'].mode()[0])

In [38]:
# Validate the change

(df.isnull().sum()/df.shape[0]) * 100

Transaction ID      0.000000
Customer ID         0.000000
Category            0.000000
Item                9.646123
Price Per Unit      4.842942
Quantity            4.803181
Total Spent         4.803181
Payment Method      0.000000
Location            0.000000
Transaction Date    0.000000
Discount Applied    0.000000
dtype: float64

##### Dropping Missing Values in Item Price Per Unit,Quantity,Total Spent columns of the dataset.

In [40]:
df.dropna(inplace=True)

In [41]:
# Validate the change

(df.isnull().sum()/df.shape[0]) * 100

Transaction ID      0.0
Customer ID         0.0
Category            0.0
Item                0.0
Price Per Unit      0.0
Quantity            0.0
Total Spent         0.0
Payment Method      0.0
Location            0.0
Transaction Date    0.0
Discount Applied    0.0
dtype: float64

### Data Cleaning

In [42]:
df.columns

Index(['Transaction ID', 'Customer ID', 'Category', 'Item', 'Price Per Unit',
       'Quantity', 'Total Spent', 'Payment Method', 'Location',
       'Transaction Date', 'Discount Applied'],
      dtype='object')

### Cleaning Category Columns

In [45]:
df['Category'].value_counts()

Category
Furniture                             1460
Electric household essentials         1437
Beverages                             1427
Food                                  1426
Milk Products                         1425
Butchers                              1421
Computers and electric accessories    1397
Patisserie                            1369
Name: count, dtype: int64